dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+  * No good place to DI it and still keep reasonably typed method references
+* Rename `ActivityContext` to just `Activity`?
+  * Mismatches Java
+  * May be strange that `Temporalio.Activities.Activity` is an instance to access with `Current` whereas
+    `Temporalio.Workflows.Workflow` is static
+
+## Workflow Definition Overview
+
+In phase 1 there was an example of an interface defining a workflow. But you can use a class instead and you can even
+have a constructor with the init params. Example:
+
+```csharp
+namespace MyNamespace;
+
+using System;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Temporalio.Activities;
+using Temporalio.Workflows;
+
+public record GreetingInfo
+{
+    [JsonPropertyName("salutation")]
+    public string Salutation { get; init; } = "Hello";
+
+    [JsonPropertyName("name")]
+    public string Name { get; init; } = "<unknown>";
+}
+
+public static class Activities
+{
+    [Activity]
+    public static string CreateGreeting(GreetingInfo info) => return $"{info.Salutation}, {info.Name}!";
+}
+
+[Workflow]
+public class GreetingWorkflow
+{
+    public static readonly GreetingWorkflow Ref = Refs.Create<GreetingWorkflow>();
+
+    private readonly BufferBlock<GreetingInfo> updates = new();
+    private readonly string name;
+    private string currentGreeting = "<unset>";
+
+    [WorkflowInit]
+    public GreetingWorkflow(string name)
+    {
+        this.name = name;
+        updates.Post(new(Name: name));
+    }
+
+    [WorkflowRun]
+    public async Task<string> RunAsync(string _)
+    {
+        // Continually update greetings while updates made available
+        while (await updates.OutputAvailableAsync())
+        {
+            // Store current greeting
+            currentGreeting = await Workflow.ExecuteActivityAsync(
+                Activities.CreateGreeting,
+                await updates.ReceiveAsync(),
+                new(StartToCloseTimeout: TimeSpan.FromSeconds(5)));
+            Workflow.Logger.LogDebug("Greeting set to {Greeting}", currentGreeting);
+        }
+        return currentGreeting;
+    }
+
+    [WorkflowSignal]
+    public Task UpdateSalutationAsync(string salutation) =>
+        updates.SendAsync(new(Salutation: salutation, Name: name));
+
+    [WorkflowSignal]
+    public Task CompleteWithGreetingAsync()
+    {
+        updates.Complete();
+        return Task.CompletedTask;
+    }
+
+    [WorkflowQuery]
+    public string CurrentGreeting() => currentGreeting;
+}
+```
+
+Some things to not about the above code (much copied from phase 1):
+
+* Every workflow must have the non-inheritable `[Workflow]` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
```

Comment:
I still maintain just having to type the name is better than the inevitable set of users that will run into this and be surprised.

Reply:
It's pretty normal to use the derived name by default (in .NET and in our SDKs). I don't think this is ever gonna matter in practice. This one gotcha is only for those that are intentionally violating .NET standards (something I've found is very rare for iface names).